src/providers/token-properties-provider.ts

```diff
@@ -0,0 +1,152 @@
+  ) {}
+
+  public async getTokensProperties(
+    tokens: Token[],
+    providerConfig?: ProviderConfig
+  ): Promise<TokenPropertiesMap> {
+    const nonAllowlistTokens = tokens.filter(
+      (token) => !this.allowList.has(token.address.toLowerCase())
+    );
+    const tokenValidationResults =
+      await this.tokenValidatorProvider.validateTokens(
+        nonAllowlistTokens,
+        providerConfig
+      );
+    const tokenToResult: TokenPropertiesMap = {};
+
+    tokens.forEach((token) => {
+      if (this.allowList.has(token.address.toLowerCase())) {
+        // if the token is in the allowlist, make it UNKNOWN so that we don't fetch the FOT fee on-chain
+        tokenToResult[token.address.toLowerCase()] = {
+          tokenValidationResult: TokenValidationResult.UNKN,
+        };
+      } else {
+        tokenToResult[token.address.toLowerCase()] = {
+          tokenValidationResult:
+            tokenValidationResults.getValidationByToken(token),
+        };
+      }
+    });
+
+    const addressesToFetchFeesOnchain: string[] = [];
+    const addressesRaw = this.buildAddressesRaw(tokens);
+
+    const tokenProperties = await this.tokenPropertiesCache.batchGet(
+      addressesRaw
+    );
+
+    // Check if we have cached token validation results for any tokens.
+    for (const address of addressesRaw) {
+      const cachedValue = tokenProperties[address];
+      if (cachedValue) {
+        tokenToResult[address] = cachedValue;
+      } else if (
+        tokenToResult[address]?.tokenValidationResult ===
+        TokenValidationResult.FOT
+      ) {
+        addressesToFetchFeesOnchain.push(address);
+      }
+    }
+
+    if (addressesToFetchFeesOnchain.length > 0) {
+      let tokenFeeMap: TokenFeeMap = {};
+
+      try {
+        tokenFeeMap = await this.tokenFeeFetcher.fetchFees(
+          addressesToFetchFeesOnchain,
+          providerConfig
+        );
+      } catch (err) {
+        log.error(
+          { err },
+          `Error fetching fees for tokens ${addressesToFetchFeesOnchain}`
+        );
+      }
+
+      await Promise.all(
+        addressesToFetchFeesOnchain.map((address) => {
+          const tokenFee = tokenFeeMap[address];
+          if (tokenFee && (tokenFee.buyFeeBps || tokenFee.sellFeeBps)) {
+            const tokenResultForAddress = tokenToResult[address];
+
+            if (tokenResultForAddress) {
+              tokenResultForAddress.tokenFeeResult = tokenFee;
+            }
+
+            // update cache concurrently
+            // at this point, we are confident that the tokens are FOT, so we can hardcode the validation result
+            return this.tokenPropertiesCache.set(
+              this.CACHE_KEY(this.chainId, address),
+              {
+                tokenFeeResult: tokenFee,
+                tokenValidationResult: TokenValidationResult.FOT,
+              }
+            );
+          } else {
+            return Promise.resolve(true);
+          }
+        })
+      );
+    }
```

Comment:
Seems like we could use `batchWrite` here: https://github.com/Uniswap/routing-api/blob/88c83324a549be81b4f8bcfdf3adaadc1bcf75c8/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L416

Reply:
I think I will have this as a separate PR follow-up.

Reply:
> Seems like we could use `batchWrite` here: https://github.com/Uniswap/routing-api/blob/88c83324a549be81b4f8bcfdf3adaadc1bcf75c8/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L416

What is the role of Ryoshi token here